### PR TITLE
HOCS-2266 : Return to private office stage from minister sign off.

### PR DESCRIPTION
**Pull request by philbarton**
Add a rejection step similar to "Not approved" but in this case go back to the PO sign-off stage.

### CODE REVIEW

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        "processes/DCU_MIN_MINISTER_SIGN_OFF.bpmn",
+        "processes/DCU_MIN_PRIVATE_OFFICE.bpmn",
+        "processes/DCU_BASE_DISPATCH.bpmn",
+        "processes/STAGE.bpmn",
+        "processes/STAGE_WITH_USER.bpmn"})
+public class DCU_MIN {
+
+    public static final String DISPATCH = "CallActivity_1rowgu5";
+    public static final String PO_SIGN_OFF = "CallActivity_0pmeblj";
+    public static final String INITIAL_DRAFT = "CallActivity_1ket68y";
```
Comment (MarkWrightBjss) [inline src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN.java:42]: This is what Will described as "smoke and mirrors" in his comment here:
https://github.com/UKHomeOffice/hocs-workflow/pull/348
I've just been renaming the workflow boxes to be more meaningful.
Do we have a better solution to Will's point?

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    }
+
+    @Test
+    public void rejectSignOffScenario() {
+
+        when(dcuMinSignOffProcess.waitsAtUserTask(APPROVE_MINISTER_SIGN_OFF))
+                .thenReturn(task -> task.complete(withVariables("valid", true, "MinisterSignOffDecision", "REJECT")));
+
+        when(dcuMinSignOffProcess.waitsAtUserTask(DCU_MIN_MinisterSignOff.VALIDATE_REJECTION_NOTE))
```
Comment (MarkWrightBjss) [inline src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java:72]: The "DCU_MIN_MinisterSignOff" qualifier here feels redundant and inconsistent.

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        verify(dcuMinSignOffProcess)
+                .hasFinished(VALIDATE_REJECTION_NOTE);
+
+        verify(dcuMinSignOffProcess)
+                .hasFinished(DCU_MIN_MINISTER_SIGN_OFF_END);
+
+        verify(dcuMinSignOffProcess, never()).hasFinished(VALIDATE_NOT_APPLICABLE);
+
+        verify(bpmnService).updateAllocationNote(any(), any(), any(), eq("REJECT"));
```
Comment (MarkWrightBjss) [inline src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java:88]: Would be good for the third "any()" here to match against the actual case note text.

**Review by MarkWrightBjss — CHANGES_REQUESTED** [review]
I think workflow unit tests come in useful when resolving conflicts in BPMN files. I might have missed it, is there a covering unit test here for the back link on the new screen?

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN.java (hunk 1 above)
Comment (philbarton) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN.java:42]: Will's point was that we're testing the workflows directly. We can't get away from that unless we change the design. My opinion is 'we are where we are' and any tests are better than non. I wasn't sure of the knock on effect of renaming them, and what would happen if there was a duplicate name?

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java (hunk 3 above)
Comment (philbarton) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java:88]: I did think that at first, but in test it comes through empty. I'll have a look and figure out how to that.

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java (hunk 2 above)
Comment (philbarton) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java:72]: Sorry, I don't understand.

Comment (MarkWrightBjss) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java:72]: when(dcuMinSignOffProcess.waitsAtUserTask(VALIDATE_REJECTION_NOTE))

Comment (philbarton) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java:72]: I see...now ta

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN.java (hunk 1 above)
Comment (MarkWrightBjss) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN.java:42]: We are where we are.
(The modeller prevents entering a duplicate.)

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java (hunk 3 above)
Comment (MarkWrightBjss) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java:88]: "VALIDATE_REJECTION_NOTE" should mock a return value for "CaseNote_MinisterReject"; that value should come through in place of the empty. (Same thing for next unit test below.)

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java (hunk 2 above)
Comment (philbarton) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java]: removed qualifire

File: src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java (hunk 3 above)
Comment (philbarton) [reply src/test/java/uk/gov/digital/ho/hocs/workflow/processes/DCU_MIN_MinisterSignOff.java]: added rejection string

**Comment by philbarton** [comment]
Also added a back link test.

**Review by MarkWrightBjss — APPROVED** [review]
Cheers, good work.

**Pull request merged**